tests/FsCheck.Test/Gen.fs

```diff
@@ -119,6 +119,12 @@ module Gen =
         Gen.listOfLength length (Gen.constant v)
         |> sample1
         |> ((=) (List.init length (fun _ -> v)))
+
+    [<Property>]
+    let Shuffle (NonEmptySet (xs:Set<int>)) =
+        Gen.shuffle xs
+        |> sample 10
+        |> List.exists ((<>) (Set.toSeq xs))
```

Comment:
> how should we verify this beast

As previously outlined, I can only come up with one property that always holds (and that is that if you sort both the original list and the shuffled list, the sorted lists should be equal to each other). Well, I can think of other properties, such as the original and shuffled list should have the same length, but such a property is 'contained' in that other property...

It's true that the property that I propose doesn't _guarantee_ that a sequence is shuffled, because a degenerate implementation could pass the test simply by returning the input sequence untouched.

All other tests I can think of rely on random behaviour. I can only think of two ways to test with randomness:

- Use deterministic random number generators. If you use a deterministic random number generator (with a constant seed), determinism is assured. Using the _Devil's Advocate_ technique, however, a malign implementer could hide in the shadows of such tests, though.
- Rely on statistics and write tests that can, but are _extraordinarily_ unlikely to, produce false positives.

If there are more alternatives, I'd love to learn them :smile: